### PR TITLE
EZP-23367: Implement variations purging

### PR DESCRIPTION
**Pull request by bdunogier**
> http://jira.ez.no/browse/EZP-23367
> Also fixes https://jira.ez.no/browse/EZP-23369

This makes it possible to remove all instances of an image alias.
## Changes
- aliases are now stored in `vardir/storage/images/_aliases/aliasName`

### Removing all variations of an alias
Can be executed after an alias has been removed from the configuration.

```
php ezpublish/console liip:imagine:cache:remove --filters=large
```

### Removing all variations
For maintenance
```
php ezpublish/console liip:imagine:cache:remove
```

### CODE REVIEW

**Comment by crevillo** [comment]
so, aliases are now save to other paths? Couldn't this have some seo issues ?

**Comment by bdunogier** [comment]
Well, for existing paths, yes, it might, that's true.

**Comment by andrerom** [comment]
@crevillo has a point as this is also an issue for 5.4, not being able to purge aliases, unless there is way of doing that using the legacy scripts, but unsure as we have a support issue on this.

**Comment by bdunogier** [comment]
> this is also an issue for 5.4, not being able to purge aliases, unless there is way of doing that using the legacy scripts, but unsure as we have a support issue on this (afaik).

New stack files can't be expired, afaik, from legacy, since they're not persisted in the same way (no database entry in ezimagefile, but @lolautruche could confirm that.

I've been thinking about 5.4 this morning, and I'm for now thinking that the path change shouldn't be backported to 5.4. It is **much** better for IO performance, including with DFS. I'd rather craft a custom solution for 5.4. What about `find var/site/storage/images -name "*_large.*" -rm` ? :-)

**Comment by crevillo** [comment]
what if the original filename is called photo_large.jpg? ;)

**Comment by bdunogier** [comment]
> what if the original filename is called photo_large.jpg? ;)

Not nice, @crevillo :p

**Comment by crevillo** [comment]
@bdunogier and how ez 5.4 knows which is the file related to an alias image? only by alias suffix? there's no relation between database and image alias?

**Comment by lolautruche** [comment]
> New stack files can't be expired, afaik, from legacy, since they're not persisted in the same way (no database entry in ezimagefile), but @lolautruche could confirm that.

I do confirm. Thoughts were that the important file is the original, an alias is just a variation of it which must be easily removable.

> and how ez 5.4 knows which is the file related to an alias image? only by alias suffix?

Yes and no. We always start by the original image name. So if the original is `image_large.jpg`, `large` alias will be `image_large_large.jpg` :wink:

**Comment by crevillo** [comment]
@lolautruche ok, but anyway @bdunogier find suggestion will delete image_large.jpg too, right? 

so, don't know how will this perform (i'm afraid badly) but thinking if we could
* query for all the image files 
* get the path of the alias for each of them
* delete that file

**Comment by crevillo** [comment]
there's something else to be taken in account imho. Imagemagick performs badly on debian installations (especially if you install from package manager). new stack won't use it but anyway, if you delete all those alias images and you have lot of pages viewed at the same time trying to re-generate needed image alias, couldn't the server be hurt?

**Comment by bdunogier** [comment]
Okay, i've thought about this a lot, and here's a solution.

With the integration of Imagine, we have decoupled image FieldType data from alias handling. But the Image FieldType still stores a reference to the original file in the `ezimagefile` table. We could:
- change the IO Repository Resolver, to get the path to ALL files in the system (with an iterator)
- use the FieldType's storage gateway (via a service) to get ALL the images
- for each path, compute the aliased path, and if it exists, remove it

The same service would exist in Platform 1.x, but wouldn't iterate over files like this, and would just remove folder (and it would be MUCH faster). Overall, it is a lot of IO, but I don't see another way out.

**Comment by joaoinacio** [comment]
My 2 cents, FWIW:
Clearing image aliases in legacy can end up being slow as it does extra checks/work (dom?), but nonetheless it has worked so far. Taken from ezcache.php:
```
        // 1. fetch ezcontentclass having an ezimage attribute
        // 2. fetch objects of these classes
        // 3. purge image alias for all version
```
The base base principle would be the same, I guess? For all content (versions) with image fields, 
Generate a list of files for all possible aliases and (try to) remove them... 

As for performance issues post-clear, wouldn't that be expected? 
Maybe some "warmup" improvement? :)

File: eZ/Bundle/EzPublishCoreBundle/Imagine/IORepositoryResolver.php [hunk 1]
```diff
@@ -134,7 +147,7 @@ public function remove( array $paths, array $filters )
      * Returns path for filtered image from original path.
      * Pattern is <original_dir>/<filename>_<filter_name>.<extension>
```
Comment (joaoinacio) [inline eZ/Bundle/EzPublishCoreBundle/Imagine/IORepositoryResolver.php]: Nitpick: should this doc comment be updated aswell?

**Comment by crevillo** [comment]
Sounds legit to me. about the warmup improvement... maybe we could delete and create it again with the new settings? it will make the thing even longer but you will be prepared to fight a visit peak... 

And well, image alias is not something we do every day, right?

**Comment by bdunogier** [comment]
> The base base principle would be the same, I guess? For all content (versions) with image fields, 

Actually no. As I (tried to) explain, we don't handle aliases the ways we used to. Aliases handling in legacy was a core part of the datatype. Now it's different. Imagine just uses the original's URI, and takes care of the rest. It won't even store aliases in the ezimagefile table, which in the end will only contain the originals.

**Comment by bdunogier** [comment]
> Sounds legit to me. about the warmup improvement... maybe we could delete and create it again with the new settings? it will make the thing even longer but you will be prepared to fight a visit peak...

Why not. I'd say it's an option, and that it can easily be added afterwards. What do you think ?

**Comment by crevillo** [comment]
:+1:

File: eZ/Bundle/EzPublishCoreBundle/Imagine/IORepositoryResolver.php (hunk 1 above)
Comment (bdunogier) [inline eZ/Bundle/EzPublishCoreBundle/Imagine/IORepositoryResolver.php]: Indeed.

**Comment by andrerom** [comment]
:+1:

**Comment by larseirik** [comment]
:+1:

**Comment by bdunogier** [comment]
Yay, plenty of thumbs.

Moving on, with a slight difference already: the service used by the `IORepositoryResolver` won't fetch the list of files to delete, it will delete the files. It is required to leverage the path change, as we don't need to list the files to delete them.

**Comment by larseirik** [comment]
Very promising! Looking forward to testing this:)

**Comment by bdunogier** [comment]
Pushed a proof-of-concept that should cover both versions.

#### 5.4, in 47063e7
Introduced a VariationPathGenerator service/interface, that returns the alias path given the original path and alias name. The 5.4 naming is implemented in the OriginalDirectory VariationPathGenerator (variation stored in the same directory than the original).

Introduced a VariationPurger service/interface, used by the IORepositoryResolver to purge all variations of one alias. Added a 5.4 implementation, ImageFileListVariationPurger. that iterates over a FileList service, that reads the contents of the ezimagefile table.

#### Platform 1.0, in 2b21688
Added a new VariationPathGenerator, AliasDirectory, that moves images to a dedicated folder per alias, in an _alias subfodler. Added a VariationPurger that uses the IOService `deleteDirectory` method to delete the alias directory.

**Comment by crevillo** [comment]
Looks good. don't know if this should go in the specs, but i'd add somewhere that you can change to the new way of "storing" this images whenever you want 
As simple as 
1. remove all aliases
2. change service

**Comment by bdunogier** [comment]
Closing in favour of #1276 (re-organized all commits for easier cherry-pick).

**Pull request closed without merging**